src/output_module.h

```diff
@@ -25,28 +25,69 @@
 #ifndef _OUTPUT_MODULE_H
 #define _OUTPUT_MODULE_H
 
+#include <set>
+#include <string>
+#include <vector>
+
 #include "output.h"
 
-struct output_module {
-  const char *shortname;
-  const char *description;
-  int (*add_options)(GOptionContext *ctx);
-
-  // Commands.
-  int (*init)(void);
-  void (*set_uri)(const char *uri, output_update_meta_cb_t meta_info);
-  void (*set_next_uri)(const char *uri);
-  int (*play)(output_transition_cb_t transition_callback);
-  int (*stop)(void);
-  int (*pause)(void);
-  int (*seek)(gint64 position_nanos);
-
-  // parameters
-  int (*get_position)(gint64 *track_duration, gint64 *track_pos);
-  int (*get_volume)(float *);
-  int (*set_volume)(float);
-  int (*get_mute)(int *);
-  int (*set_mute)(int);
+class OutputModule {
+ public:
+  struct Options {
+    virtual std::vector<GOptionGroup*> GetOptionGroups(void) = 0;
+  };
+
+  struct TrackState {
+    int64_t duration_ns;
+    int64_t position_ns;
+  };
+
+  typedef enum Result { kSuccess = 0, kError = -1 } Result;
+
+  OutputModule(Output::PlaybackCallback play = nullptr,
+               Output::MetadataCallback meta = nullptr)
+      : playback_callback_(play), metadata_callback_(meta) {}
+
+  virtual Result Initalize(Options& options) = 0;
+
+  virtual Output::MimeTypeSet GetSupportedMedia(void) = 0;
+
+  virtual void SetUri(const std::string& uri) = 0;
+  virtual void SetNextUri(const std::string& uri) = 0;
+
+  virtual Result Play(void) = 0;
+  virtual Result Stop(void) = 0;
+  virtual Result Pause(void) = 0;
+  virtual Result Seek(int64_t position_ns) = 0;
+
+  virtual Result GetPosition(TrackState& track) = 0;
+  virtual Result GetVolume(float& volume) = 0;
```

Comment:
The return parameters should all be pointers `GetVolume(float *volume)`
(discussed above).

Reply:
Another question is: do we actually want the Result thing in these cases ? I guess the question is if we ever want to indicate that we don't support something, but we could just simplify everything if we consider
```
   virtual void SetVolume(float v);
   virtual float GetVolume();
```

Might be something to consider for some re-condiering cleanup change later though.

Reply:
There are a few locations that check the result. Another option could be to return a struct/pair with a result indication and the value.